cgen error pointer expected

### Describe the bug

encountered cgen error when rewriting some tests

### Reproduction Steps

```bash
git clone https://github.com/einar-hjortdal/firebird.git
cd firebird
git checkout 11566658f6fdecef0ac7d5c3f0bb5ed1eb28e6d2
v install
v run statement_test.v
```

### Expected Behavior

no cgen error

### Current Behavior

```
v run statement_test.v 
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/statement_test.01KT45TRMJEQNJ27NSMC34E3W5.tmp.c:26140: warning: integer constant overflow
cc: /tmp/v_1000/statement_test.01KT45TRMJEQNJ27NSMC34E3W5.tmp.c:35276: error: pointer expected
=================================================================
Try passing `-g` when compiling, to see a .v file:line information, that correlates more with the C error.
(Alternatively, pass `-show-c-output`, to print the full C error message).
builder error: 
==================
C error found while compiling generated C code.
This can be caused by invalid C interop code, C compiler flags, or a V compiler bug.
If your code is pure V and this still happens, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.5.1 7bf483e

### Environment details (OS name and version, etc.)

|V full version      |V 0.5.1 58f18573557247e9c743f7d60e5b3ab783f74479.7bf483e
|:-------------------|:-------------------
|OS                  |linux, "openmamba release 2026.4.1 for x86_64 (rolling)"
|Processor           |8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
|Memory              |14.65GB/31.17GB
|                    |
|V executable        |/home/einar/.local/lib64/v/v
|V last modified time|2026-06-02 11:28:21
|                    |
|V home dir          |OK, value: /home/einar/.local/lib64/v
|VMODULES            |OK, value: /home/einar/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/einar/Documents/projects/vlang/active/firebird
|                    |
|Git version         |git version 2.54.0
|V git status        |weekly.2026.03-2194-g7bf483e2
|.git/config present |true
|                    |
|cc version          |cc (GCC) 16.1.0
|gcc version         |gcc (GCC) 16.1.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.43
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/scope.v
+++ b/vlib/v/ast/scope.v
@@ -112,6 +112,35 @@ pub fn (s &Scope) find_var(name string) ?&Var {
 	return none
 }
 
+// find_var_decl returns the nearest declaration variable named `name`, walking
+// the parent scope chain and skipping synthetic smartcast vars (introduced by
+// `if x is T`/`match` branches, which carry a non-empty `smartcasts` list and
+// are copies of the real declaration). Use this when a promotion such as
+// auto-heap must be recorded on the variable that codegen emits as the
+// declaration, not on a smartcast copy living in an inner branch scope.
+pub fn (s &Scope) find_var_decl(name string) ?&Var {
+	if _unlikely_(s == unsafe { nil }) {
+		return none
+	}
+	for sc := unsafe { s }; true; sc = sc.parent {
+		pobj := unsafe { &sc.objects[name] or { nil } }
+		if pobj != unsafe { nil } {
+			match pobj {
+				Var {
+					if pobj.smartcasts.len == 0 {
+						return &pobj
+					}
+				}
+				else {}
+			}
+		}
+		if sc.dont_lookup_parent() {
+			break
+		}
+	}
+	return none
+}
+
 pub fn (s &Scope) find_global(name string) ?&GlobalField {
 	obj := s.find_ptr(name)
 	if _likely_(obj != unsafe { nil }) {

--- a/vlib/v/ast/scope.v
+++ b/vlib/v/ast/scope.v
@@ -113,11 +113,17 @@ pub fn (s &Scope) find_var(name string) ?&Var {
 }
 
 // find_var_decl returns the nearest declaration variable named `name`, walking
-// the parent scope chain and skipping synthetic smartcast vars (introduced by
-// `if x is T`/`match` branches, which carry a non-empty `smartcasts` list and
-// are copies of the real declaration). Use this when a promotion such as
-// auto-heap must be recorded on the variable that codegen emits as the
-// declaration, not on a smartcast copy living in an inner branch scope.
+// the parent scope chain and skipping synthetic smartcast vars introduced by
+// `if x is T`/`match` branches. Use this when a promotion such as auto-heap must
+// be recorded on the variable that codegen emits as the declaration, not on a
+// smartcast copy living in an inner branch scope.
+//
+// Only true synthetic copies are skipped: those are registered as separate
+// scope objects carrying both a non-empty `smartcasts` list and a non-zero
+// `orig_type` (the pre-smartcast type). A real declaration that is smartcast in
+// place keeps `orig_type == 0` (e.g. an option var unwrapped by
+// `if x == none { continue }` via `update_smartcasts`), so it is returned
+// rather than skipped.
 pub fn (s &Scope) find_var_decl(name string) ?&Var {
 	if _unlikely_(s == unsafe { nil }) {
 		return none
@@ -127,7 +133,7 @@ pub fn (s &Scope) find_var_decl(name string) ?&Var {
 		if pobj != unsafe { nil } {
 			match pobj {
 				Var {
-					if pobj.smartcasts.len == 0 {
+					if pobj.smartcasts.len == 0 || pobj.orig_type == 0 {
 						return &pobj
 					}
 				}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -7646,13 +7646,21 @@ fn (mut c Checker) mark_as_referenced(mut node ast.Expr, as_interface bool) {
 				// use-site scope chain, since `c.fn_scope.find_var` only locates
 				// variables declared at the function level and misses those
 				// declared in nested scopes (e.g. inside a `for` loop body).
-				obj = node.scope.find_var(node.obj.name) or {
-					if c.fn_scope != unsafe { nil } {
+				mut resolved := node.scope.find_var(node.obj.name) or { obj }
+				if resolved.smartcasts.len != 0 {
+					// `node.scope.find_var` resolved a synthetic smartcast var
+					// from an `if x is T`/`match` branch rather than the real
+					// declaration. Such vars have no declaration of their own, so
+					// promoting them would set `is_auto_heap` on the smartcast var
+					// while the actual declaration stays a plain value, desyncing
+					// the pointer indirection. Fall back to the function scope.
+					resolved = if c.fn_scope != unsafe { nil } {
 						c.fn_scope.find_var(node.obj.name) or { obj }
 					} else {
 						obj
 					}
 				}
+				obj = resolved
 				if obj.typ == 0 {
 					return
 				}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -7695,6 +7695,21 @@ fn (mut c Checker) mark_as_referenced(mut node ast.Expr, as_interface bool) {
 						}
 					}
 
+					if obj.is_auto_heap {
+						// `obj` is the canonical declaration, which cgen emits as a
+						// heap pointer. When the value is read through a branch-local
+						// smartcast/option-unwrap copy (`if x is T`, `if x != none`),
+						// cgen resolves that use-site copy (see
+						// `resolved_ident_is_auto_heap`) when emitting the read, so it
+						// must carry the same `is_auto_heap`. Otherwise the read (e.g.
+						// an unwrapped option's `.data`) is emitted without the pointer
+						// indirection the declaration was given, producing invalid C.
+						node.obj.is_auto_heap = true
+						if mut use_site := node.scope.find_var(node.obj.name) {
+							use_site.is_auto_heap = true
+						}
+					}
+
 					if as_interface {
 						for method in type_sym.methods {
 							c.mark_fn_decl_as_referenced(method.fkey())

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -7640,8 +7640,18 @@ fn (mut c Checker) mark_as_referenced(mut node ast.Expr, as_interface bool) {
 		ast.Ident {
 			if mut node.obj is ast.Var {
 				mut obj := unsafe { &node.obj }
-				if c.fn_scope != unsafe { nil } {
-					obj = c.fn_scope.find_var(node.obj.name) or { obj }
+				// Resolve the canonical declaration variable so that the
+				// `is_auto_heap` promotion below is recorded on the variable
+				// that cgen will see when emitting its declaration. Prefer the
+				// use-site scope chain, since `c.fn_scope.find_var` only locates
+				// variables declared at the function level and misses those
+				// declared in nested scopes (e.g. inside a `for` loop body).
+				obj = node.scope.find_var(node.obj.name) or {
+					if c.fn_scope != unsafe { nil } {
+						c.fn_scope.find_var(node.obj.name) or { obj }
+					} else {
+						obj
+					}
 				}
 				if obj.typ == 0 {
 					return

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -7642,25 +7642,21 @@ fn (mut c Checker) mark_as_referenced(mut node ast.Expr, as_interface bool) {
 				mut obj := unsafe { &node.obj }
 				// Resolve the canonical declaration variable so that the
 				// `is_auto_heap` promotion below is recorded on the variable
-				// that cgen will see when emitting its declaration. Prefer the
-				// use-site scope chain, since `c.fn_scope.find_var` only locates
-				// variables declared at the function level and misses those
-				// declared in nested scopes (e.g. inside a `for` loop body).
-				mut resolved := node.scope.find_var(node.obj.name) or { obj }
-				if resolved.smartcasts.len != 0 {
-					// `node.scope.find_var` resolved a synthetic smartcast var
-					// from an `if x is T`/`match` branch rather than the real
-					// declaration. Such vars have no declaration of their own, so
-					// promoting them would set `is_auto_heap` on the smartcast var
-					// while the actual declaration stays a plain value, desyncing
-					// the pointer indirection. Fall back to the function scope.
-					resolved = if c.fn_scope != unsafe { nil } {
+				// that cgen will see when emitting its declaration. Walk the
+				// use-site scope chain (so variables declared in nested scopes,
+				// e.g. inside a `for` loop body, are found, unlike
+				// `c.fn_scope.find_var` which only locates function-level vars),
+				// while skipping synthetic smartcast vars from `if x is T`/`match`
+				// branches: those are copies with no declaration of their own, so
+				// promoting them would leave the real declaration emitted as a
+				// plain value and desync the pointer indirection.
+				obj = node.scope.find_var_decl(node.obj.name) or {
+					if c.fn_scope != unsafe { nil } {
 						c.fn_scope.find_var(node.obj.name) or { obj }
 					} else {
 						obj
 					}
 				}
-				obj = resolved
 				if obj.typ == 0 {
 					return
 				}

--- a/vlib/v/tests/interfaces/interface_variadic_test.v
+++ b/vlib/v/tests/interfaces/interface_variadic_test.v
@@ -101,6 +101,33 @@ fn test_variadic_interface_arg_smartcast_in_loop() {
 	assert total == 5
 }
 
+fn maybe_value() ?Cat {
+	return Cat{}
+}
+
+// For issue 27326: an option value unwrapped via an if-guard or `if x != none`
+// and then passed to a variadic interface parameter is auto-heap promoted, so
+// its declaration is emitted as an option pointer. The unwrapped `.data` read
+// at the call site must use the matching pointer indirection (`->data`);
+// previously cgen emitted value-style access and failed C compilation.
+fn test_variadic_interface_arg_option_unwrap() {
+	mut total := 0
+	if x := maybe_value() {
+		total += collect(x)
+	}
+	x := maybe_value()
+	if x != none {
+		total += collect(x)
+	}
+	for _ in 0 .. 2 {
+		y := maybe_value()
+		if y != none {
+			total += collect(y)
+		}
+	}
+	assert total == 4
+}
+
 fn check_animals(animals ...Animal) {
 	assert animals[0] is Cat
 	assert animals[1] is Dog

--- a/vlib/v/tests/interfaces/interface_variadic_test.v
+++ b/vlib/v/tests/interfaces/interface_variadic_test.v
@@ -36,6 +36,26 @@ fn test_variadic_interface_fn_arg() {
 	check_animals(c, d)
 }
 
+// For issue 27326: passing a local variable declared inside a loop to a
+// variadic interface parameter caused a `pointer expected` C error, because
+// the value was promoted to auto-heap at the use site but not at its
+// declaration, producing inconsistent pointer indirection in the generated C.
+interface Value {}
+
+fn collect(params ...Value) int {
+	return params.len
+}
+
+fn test_variadic_interface_arg_declared_in_loop() {
+	mut total := 0
+	for i := 0; i < 3; i++ {
+		id := [u8(1), 2, 3]
+		code := 'hello ${i}'
+		total += collect(id, code)
+	}
+	assert total == 6
+}
+
 fn check_animals(animals ...Animal) {
 	assert animals[0] is Cat
 	assert animals[1] is Dog

--- a/vlib/v/tests/interfaces/interface_variadic_test.v
+++ b/vlib/v/tests/interfaces/interface_variadic_test.v
@@ -56,6 +56,30 @@ fn test_variadic_interface_arg_declared_in_loop() {
 	assert total == 6
 }
 
+type ValueSum = Cat | Dog
+
+// For issue 27326: a variable smartcast via `if x is T` / `match` resolves to a
+// synthetic smartcast scope var, not its real declaration. Promoting that
+// synthetic var to auto-heap instead of the declaration must not desync the
+// pointer indirection when the value is passed to a variadic interface param.
+fn test_variadic_interface_arg_smartcast() {
+	mut total := 0
+	s := ValueSum(Cat{})
+	if s is Cat {
+		total += collect(s)
+	}
+	v := Value(Cat{})
+	if v is Cat {
+		total += collect(v)
+	}
+	match v {
+		Cat { total += collect(v) }
+		else {}
+	}
+
+	assert total == 3
+}
+
 fn check_animals(animals ...Animal) {
 	assert animals[0] is Cat
 	assert animals[1] is Dog

--- a/vlib/v/tests/interfaces/interface_variadic_test.v
+++ b/vlib/v/tests/interfaces/interface_variadic_test.v
@@ -80,6 +80,27 @@ fn test_variadic_interface_arg_smartcast() {
 	assert total == 3
 }
 
+// For issue 27326: a variable that is both declared in a nested scope (a `for`
+// loop body) and smartcast via `if x is T` must still resolve to its real
+// nested declaration, not the synthetic smartcast var nor only the function
+// scope, otherwise the auto-heap promotion desyncs the pointer indirection.
+fn test_variadic_interface_arg_smartcast_in_loop() {
+	mut total := 0
+	for _ in 0 .. 3 {
+		s := ValueSum(Cat{})
+		if s is Cat {
+			total += collect(s)
+		}
+	}
+	for _ in 0 .. 2 {
+		v := Value(Cat{})
+		if v is Cat {
+			total += collect(v)
+		}
+	}
+	assert total == 5
+}
+
 fn check_animals(animals ...Animal) {
 	assert animals[0] is Cat
 	assert animals[1] is Dog

